module/zfs/dmu_objset.c

```diff
@@ -973,16 +1030,33 @@ dmu_objset_create(const char *name, dmu_objset_type_t type, uint64_t flags,
 	doca.doca_userfunc = func;
 	doca.doca_userarg = arg;
 	doca.doca_type = type;
+	doca.doca_dcp = dcp;
 
-	return (dsl_sync_task(name,
+	ret = dsl_sync_task(name,
 	    dmu_objset_create_check, dmu_objset_create_sync, &doca,
-	    5, ZFS_SPACE_CHECK_NORMAL));
+	    5, ZFS_SPACE_CHECK_NORMAL);
+	if (ret != 0)
+		return (ret);
+
+	ret = dmu_objset_hold(name, FTAG, &os);
+	if (ret != 0)
+		return (ret);
+
+	/* See comment in dmu_objset_create_sync() for details */
+	if (os->os_encrypted)
+		(void) spa_keystore_remove_mapping(os->os_spa,
+		    os->os_dsl_dataset, &doca);
```

Comment:
> This seems best-effort and not guaranteed to work, since another thread could have e.g renamed the dataset after the sync task completes. 

I'm not quite sure how a rename or anything else could interfere with this. This should always work. Can you elaborate?

>  But it seems like this is required because it's releasing a hold that was put on by the sync func?

Correct, see my comment above.